### PR TITLE
AB#87737 Added Milestone dates validation

### PR DESCRIPTION
**Pull request by Antek012**

### CODE REVIEW

File: src/HE.Investment.AHP.Domain/Delivery/CommandHandlers/CreateDeliveryPhaseCommandHandler.cs [hunk 1]
```diff
@@ -31,9 +37,9 @@ public class CreateDeliveryPhaseCommandHandler : DeliveryCommandHandlerBase, IRe
         try
         {
             var account = await _accountUserContext.GetSelectedAccount();
-
+            var programme = await _programmeRepository.GetCurrentProgramme(ProgrammeType.Ahp);
```
Comment (jjankowski87) [inline src/HE.Investment.AHP.Domain/Delivery/CommandHandlers/CreateDeliveryPhaseCommandHandler.cs:40]: Create GetProgrammeBasicInfo method in ProgrammeRepository and use it in DeliveryPhaseRepository as well

**Review by jjankowski87 — APPROVED** [review]

File: src/HE.Investment.AHP.Domain/Delivery/CommandHandlers/CreateDeliveryPhaseCommandHandler.cs (hunk 1 above)
Comment (Antek012) [reply src/HE.Investment.AHP.Domain/Delivery/CommandHandlers/CreateDeliveryPhaseCommandHandler.cs:40]: good idea - changed

**Review by jjankowski87 — APPROVED** [review]

File: src/HE.Investment.AHP.Domain.Tests/Delivery/Entities/TestDataBuilders/MilestonePaymentDateFactory.cs [hunk 2]
```diff
@@ -0,0 +1,20 @@
+using System.Globalization;
+using HE.Investment.AHP.Domain.Delivery.ValueObjects;
+
+namespace HE.Investment.AHP.Domain.Tests.Delivery.Entities.TestDataBuilders;
+
+public class MilestonePaymentDateFactory
+{
+    public static MilestonePaymentDate GetDateDayAfter(MilestonePaymentDate? date)
+    {
+        if (date == null || date.Value.Day == 28)
```
Comment (Carq) [inline src/HE.Investment.AHP.Domain.Tests/Delivery/Entities/TestDataBuilders/MilestonePaymentDateFactory.cs:10]: Why 28 is a special number? Could you put value into well named variable?

File: src/HE.Investment.AHP.Domain/Delivery/CommandHandlers/ProvideAcquisitionMilestoneDetailsCommandHandler.cs [hunk 3]
```diff
@@ -22,9 +22,11 @@ protected override Task<OperationResult> Update(IDeliveryPhaseEntity entity, Pro
             AcquisitionDate.Create(request.AcquisitionDate.Day, request.AcquisitionDate.Month, request.AcquisitionDate.Year));
         var milestonePaymentDate = operationResult.AggregateNullable(() =>
             MilestonePaymentDate.Create(request.PaymentDate.Day, request.PaymentDate.Month, request.PaymentDate.Year));
-        var details = operationResult.AggregateNullable(() => AcquisitionMilestoneDetails.Create(acquisitionDate, milestonePaymentDate));
+        var milestone = operationResult.AggregateNullable(() => AcquisitionMilestoneDetails.Create(acquisitionDate, milestonePaymentDate, entity.DeliveryPhaseMilestones.Programme));
```
Comment (Carq) [inline src/HE.Investment.AHP.Domain/Delivery/CommandHandlers/ProvideAcquisitionMilestoneDetailsCommandHandler.cs:25]: Liskov rule.

File: src/HE.Investment.AHP.Domain/Delivery/Entities/DeliveryPhasesEntity.cs [hunk 4]
```diff
@@ -85,14 +86,17 @@ public void SetHomesToBeDeliveredInPhase(DeliveryPhaseId deliveryPhaseId, IReadO
         }
     }
 
-    public DeliveryPhaseEntity CreateDeliveryPhase(DeliveryPhaseName name)
+    public DeliveryPhaseEntity CreateDeliveryPhase(DeliveryPhaseName name, ProgrammeBasicInfo programmeBasicInfo)
```
Comment (Carq) [inline src/HE.Investment.AHP.Domain/Delivery/Entities/DeliveryPhasesEntity.cs:89]: Do we need to pass `ProgrammeBasicInfo` here?

File: src/HE.Investment.AHP.Domain/Delivery/Entities/DeliveryPhasesEntity.cs [hunk 5]
```diff
@@ -102,7 +106,8 @@ public DeliveryPhaseEntity CreateDeliveryPhase(DeliveryPhaseName name)
             null,
             new BuildActivityType(),
             SectionStatus.InProgress,
-            new List<HomesToDeliverInPhase>());
+            new List<HomesToDeliverInPhase>(),
```
Comment (Carq) [inline src/HE.Investment.AHP.Domain/Delivery/Entities/DeliveryPhasesEntity.cs:109]: User Array.Empty method.

File: src/HE.Investment.AHP.Domain/Delivery/ValueObjects/MilestoneDetails.cs [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    protected override IEnumerable<object?> GetAtomicValues()
+    {
+        yield return MilestoneDate;
+        yield return PaymentDate;
+    }
+
+    private static void ValidateDateInProgrammeRange(DateValueObject? date, ProgrammeBasicInfo programme, string fieldName, string fieldLabel)
+    {
+        if (date != null &&
+            (date.Value < programme.StartAt || date.Value > programme.EndAt))
```
Comment (Carq) [inline src/HE.Investment.AHP.Domain/Delivery/ValueObjects/MilestoneDetails.cs:46]: I think that this logic could be inside DateValueObject

File: src/HE.Investment.AHP.Domain/HE.Investment.AHP.Domain.csproj [hunk 7]
```diff
@@ -46,6 +46,10 @@
     </Reference>
   </ItemGroup>
 
+  <ItemGroup>
```
Comment (Carq) [inline src/HE.Investment.AHP.Domain/HE.Investment.AHP.Domain.csproj:49]: Is it needed?

**Review by Carq — APPROVED** [review]

File: src/HE.Investment.AHP.Domain.Tests/Delivery/Entities/TestDataBuilders/MilestonePaymentDateFactory.cs (hunk 2 above)
Comment (Antek012) [reply src/HE.Investment.AHP.Domain.Tests/Delivery/Entities/TestDataBuilders/MilestonePaymentDateFactory.cs]: changed

File: src/HE.Investment.AHP.Domain/Delivery/Entities/DeliveryPhasesEntity.cs (hunk 4 above)
Comment (Antek012) [reply src/HE.Investment.AHP.Domain/Delivery/Entities/DeliveryPhasesEntity.cs]: changed

File: src/HE.Investment.AHP.Domain/HE.Investment.AHP.Domain.csproj (hunk 7 above)
Comment (Antek012) [reply src/HE.Investment.AHP.Domain/HE.Investment.AHP.Domain.csproj]: removed

File: src/HE.Investment.AHP.Domain/Delivery/ValueObjects/MilestoneDetails.cs (hunk 6 above)
Comment (Antek012) [reply src/HE.Investment.AHP.Domain/Delivery/ValueObjects/MilestoneDetails.cs]: changed

File: src/HE.Investment.AHP.Domain/Delivery/CommandHandlers/ProvideAcquisitionMilestoneDetailsCommandHandler.cs (hunk 3 above)
Comment (Antek012) [reply src/HE.Investment.AHP.Domain/Delivery/CommandHandlers/ProvideAcquisitionMilestoneDetailsCommandHandler.cs]: changed

**Pull request merged**